local folder: open default filemanager when clicking on View...

wdlkmpx opened the pull request:
[local] Right Click on a directory -> View...

wdlkmpx commented:
I added `do not view remote folders` https://github.com/masneyb/gftp/pull/18/commits/395ba674998d738107e7dafbc08bbe2b78b48b1c

It works ok in my tests

wdlkmpx commented:
I see this patch is.. wrong. I can edit this PR in many different ways.

The gftp viewer is also defective, it opens tar files in the viewer.

gftp should detect if it's a text file and open the file in the internal viewer otherwise the program to call is xdg-open with the file/directory path.

But I think xdg-open is linux-specific I don't know. 
https://linux.die.net/man/1/xdg-open

Also reading some gmrun patches... glib also has some functions to retrieve the preferred app or something..

wdlkmpx commented:
I think the PR is ok now. I can also make gftp open only text files with the internal text viewer, but that will happen after this pr is merged. Too many branches, possible merge conflicts in the horizon.

masneyb commented on src/gtk/view_dialog.c:
```diff
@@ -44,8 +44,20 @@ do_view_or_edit_file (gftp_window_data * fromwdata, int is_view)
+        if (strcmp (gftp_protocols[fromwdata->request->protonum].name, "Local") == 0)
+        {
+          char cmd[2048]={'\0'};
+          sprintf(cmd,"xdg-open '%s'",curfle->file);
```
Use `g_snprintf` instead of `sprintf` and you can remove the NUL byte for cmd above.

masneyb replied:
Actually, you don't want to do that. You need to use something like `execve` instead. fork_process() in src/gtk/view_dialog.c has an example. Really that function should be split up and parts moved into lib/.

masneyb replied:
I also suggest checking to see if glib has some functionality to do this since that may make it easier and more secure.

masneyb commented on src/gtk/view_dialog.c:
```diff
@@ -44,8 +44,20 @@ do_view_or_edit_file (gftp_window_data * fromwdata, int is_view)
+        {
+          char cmd[2048]={'\0'};
+          sprintf(cmd,"xdg-open '%s'",curfle->file);
+          ftp_log (gftp_logging_error, NULL, _("View directory: run %s\n"), cmd);
```
maybe use level `gftp_logging_misc` instead of `gftp_logging_error`

masneyb commented on src/gtk/view_dialog.c:
```diff
@@ -44,8 +44,20 @@ do_view_or_edit_file (gftp_window_data * fromwdata, int is_view)
-        ftp_log (gftp_logging_error, NULL,
-                 _("View: %s is a directory. Cannot view it.\n"), curfle->file);
+      {
+        if (strcmp (gftp_protocols[fromwdata->request->protonum].name, "Local") == 0)
```
This string comparison is fine since that's how the code is currently structured. If you are looking for other improvements to gftp, then I suggest adding to the supported_gftp_protocols structure a new enum field for each of the supported protocols. Right now, if there's a slight typo in the name then the compiler then the compiler won't break the build.

wdlkmpx commented:
I think I addressed all the observations:
- `g_snprintf` instead of `sprintf`
- use `gftp_logging_misc` for `View directory: run...`
- use glib to run the command.

Several projects use `g_spawn_command_line_async` which works fine actually.

I also changed cmd[2048] to cmd[512], I think the buffer should be smaller.
